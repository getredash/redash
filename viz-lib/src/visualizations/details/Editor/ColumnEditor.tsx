import { map } from "lodash";
import React from "react";
import { useDebouncedCallback } from "use-debounce";
import * as Grid from "antd/lib/grid";
import { Section, Select, Input, TextAlignmentSelect } from "@/components/visualizations/editor";

import ColumnTypes from "../../table/columns";

type OwnProps = {
  column: {
    name: string;
    title?: string;
    visible?: boolean;
    alignContent?: "left" | "center" | "right";
    displayAs?: any;
    description?: string;
  };
  onChange?: (...args: any[]) => any;
};

type Props = OwnProps & typeof ColumnEditor.defaultProps;

export default function ColumnEditor({ column, onChange }: Props) {
  function handleChange(changes: any) {
    onChange({ ...column, ...changes });
  }

  const [handleChangeDebounced] = useDebouncedCallback(handleChange, 200);

  // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
  const AdditionalOptions = ColumnTypes[column.displayAs].Editor || null;

  return (
    <div className="details-visualization-editor-column">
      {/* @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message */}
      <Section>
        {/* @ts-expect-error ts-migrate(2322) FIXME: Type '{ children: Element[]; gutter: number; type:... Remove this comment to see the full error message */}
        <Grid.Row gutter={15} type="flex" align="middle">
          <Grid.Col span={16}>
            <Input
              label="Column Title"
              data-test={`Details.Column.${column.name}.Title`}
              defaultValue={column.title}
              onChange={(event: any) => handleChangeDebounced({ title: event.target.value })}
            />
          </Grid.Col>
          <Grid.Col span={8}>
            <TextAlignmentSelect
              label="Alignment"
              data-test={`Details.Column.${column.name}.TextAlignment`}
              defaultValue={column.alignContent}
              onChange={(event: any) => handleChange({ alignContent: event.target.value })}
            />
          </Grid.Col>
        </Grid.Row>
      </Section>

      {/* @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message */}
      <Section>
        <Input
          label="Description"
          data-test={`Details.Column.${column.name}.Description`}
          defaultValue={column.description}
          onChange={(event: any) => handleChangeDebounced({ description: event.target.value })}
        />
      </Section>

      {/* @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message */}
      <Section>
        <Select
          label="Display as:"
          data-test={`Details.Column.${column.name}.DisplayAs`}
          defaultValue={column.displayAs}
          onChange={(displayAs: any) => handleChange({ displayAs })}>
          {map(ColumnTypes, ({ friendlyName }, key) => (
            // @ts-expect-error ts-migrate(2339) FIXME: Property 'Option' does not exist on type '({ class... Remove this comment to see the full error message
            <Select.Option key={key} data-test={`Details.Column.${column.name}.DisplayAs.${key}`}>
              {friendlyName}
              {/* @ts-expect-error ts-migrate(2339) FIXME: Property 'Option' does not exist on type '({ class... Remove this comment to see the full error message */}
            </Select.Option>
          ))}
        </Select>
      </Section>

      {AdditionalOptions && <AdditionalOptions column={column} onChange={handleChange} />}
    </div>
  );
}

ColumnEditor.defaultProps = {
  onChange: (...args: any[]) => {},
};
