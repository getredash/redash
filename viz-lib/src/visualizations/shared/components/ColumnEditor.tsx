import { map } from "lodash";
import React from "react";
import { useDebouncedCallback } from "use-debounce";
import * as Grid from "antd/lib/grid";
import { Section, Select, Input, Checkbox, TextAlignmentSelect } from "@/components/visualizations/editor";

import ColumnTypes from "../columns";

type Column = {
  name: string;
  title?: string;
  visible?: boolean;
  alignContent?: "left" | "center" | "right";
  displayAs?: any;
  description?: string;
  allowSearch?: boolean;
};

type ColumnEditorProps = {
  column: Column;
  onChange?: (changes: any) => any;
  variant: "table" | "details";
  showSearch?: boolean;
  testPrefix?: string;
};

export default function ColumnEditor({
  column,
  onChange,
  variant,
  showSearch = variant === "table",
  testPrefix,
}: ColumnEditorProps) {
  function handleChange(changes: any) {
    if (onChange) {
      onChange({ ...column, ...changes });
    }
  }

  const [handleChangeDebounced] = useDebouncedCallback(handleChange, 200);

  // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
  const AdditionalOptions = ColumnTypes[column.displayAs].Editor || null;

  const cssClass = `${variant}-visualization-editor-column`;
  const dataTestPrefix = testPrefix || `${variant === "table" ? "Table" : "Details"}.Column.${column.name}`;

  return (
    <div className={cssClass}>
      {/* @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message */}
      <Section>
        {/* @ts-expect-error ts-migrate(2322) FIXME: Type '{ children: Element[]; gutter: number; type:... Remove this comment to see the full error message */}
        <Grid.Row gutter={15} type="flex" align="middle">
          <Grid.Col span={16}>
            <Input
              data-test={`${dataTestPrefix}.Title`}
              defaultValue={column.title}
              onChange={(event: any) => handleChangeDebounced({ title: event.target.value })}
            />
          </Grid.Col>
          <Grid.Col span={8}>
            <TextAlignmentSelect
              data-test={`${dataTestPrefix}.TextAlignment`}
              defaultValue={column.alignContent}
              onChange={(event: any) => handleChange({ alignContent: event.target.value })}
            />
          </Grid.Col>
        </Grid.Row>
      </Section>

      {showSearch && (
        /* @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message */
        <Section>
          <Checkbox
            data-test={`${dataTestPrefix}.UseForSearch`}
            defaultChecked={column.allowSearch}
            onChange={event => handleChange({ allowSearch: event.target.checked })}>
            Use for search
          </Checkbox>
        </Section>
      )}

      {/* @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message */}
      <Section>
        <Input
          label="Description"
          data-test={`${dataTestPrefix}.Description`}
          defaultValue={column.description}
          onChange={(event: any) => handleChangeDebounced({ description: event.target.value })}
        />
      </Section>

      {/* @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message */}
      <Section>
        <Select
          label="Display as:"
          data-test={`${dataTestPrefix}.DisplayAs`}
          defaultValue={column.displayAs}
          onChange={(displayAs: any) => handleChange({ displayAs })}>
          {map(ColumnTypes, ({ friendlyName }, key) => (
            // @ts-expect-error ts-migrate(2339) FIXME: Property 'Option' does not exist on type '({ class... Remove this comment to see the full error message
            <Select.Option key={key} data-test={`${dataTestPrefix}.DisplayAs.${key}`}>
              {friendlyName}
              {/* @ts-expect-error ts-migrate(2339) FIXME: Property 'Option' does not exist on type '({ class... Remove this comment to see the full error message */}
            </Select.Option>
          ))}
        </Select>
      </Section>

      {AdditionalOptions && <AdditionalOptions column={column} onChange={handleChange} />}
    </div>
  );
}

ColumnEditor.defaultProps = {
  onChange: () => {},
};
