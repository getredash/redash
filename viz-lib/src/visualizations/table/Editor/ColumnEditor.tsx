import React from "react";
import SharedColumnEditor from "../../shared/components/ColumnEditor";

type OwnProps = {
  column: {
    name: string;
    title?: string;
    visible?: boolean;
    alignContent?: "left" | "center" | "right";
    displayAs?: any;
    allowSearch?: boolean;
    description?: string;
  };
  onChange?: (...args: any[]) => any;
};

type Props = OwnProps & typeof ColumnEditor.defaultProps;

export default function ColumnEditor({ column, onChange }: Props) {
  return (
    <SharedColumnEditor
      column={column}
      onChange={onChange}
      variant="table"
      showSearch={true}
    />
  );
}

ColumnEditor.defaultProps = {
  onChange: () => {},
};
