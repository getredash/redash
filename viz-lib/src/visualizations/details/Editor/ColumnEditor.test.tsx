import React from "react";
import enzyme from "enzyme";

import ColumnEditor from "./ColumnEditor";

function findByTestID(wrapper: any, testId: any) {
  return wrapper.find(`[data-test="${testId}"]`);
}

function mount(column: any, done: any) {
  return enzyme.mount(
    <ColumnEditor
      column={column}
      onChange={(changedColumn: any) => {
        expect(changedColumn).toMatchSnapshot();
        done();
      }}
    />
  );
}

describe("Visualizations -> Details -> Editor -> Column Editor", () => {
  test("Changes column title", done => {
    const el = mount(
      {
        name: "user_id",
        title: "user_id",
        visible: true,
        alignContent: "left",
        displayAs: "string",
        description: "",
      },
      done
    );

    findByTestID(el, "Details.Column.user_id.Title")
      .last()
      .simulate("change", { target: { value: "User ID" } });
  });

  test("Changes column alignment", done => {
    const el = mount(
      {
        name: "amount",
        title: "amount",
        visible: true,
        alignContent: "left",
        displayAs: "number",
        description: "",
      },
      done
    );

    findByTestID(el, "Details.Column.amount.TextAlignment")
      .last()
      .find('input[value="right"]')
      .simulate("change", { target: { value: "right" } });
  });

  test("Changes column description", done => {
    const el = mount(
      {
        name: "status",
        title: "Status",
        visible: true,
        alignContent: "left",
        displayAs: "string",
        description: "",
      },
      done
    );

    findByTestID(el, "Details.Column.status.Description")
      .last()
      .simulate("change", { target: { value: "Current order status" } });
  });

  test("Changes display type", done => {
    const el = mount(
      {
        name: "created_at",
        title: "Created At",
        visible: true,
        alignContent: "left",
        displayAs: "datetime",
        description: "",
      },
      done
    );

    findByTestID(el, "Details.Column.created_at.DisplayAs")
      .last()
      .simulate("mouseDown");
    findByTestID(el, "Details.Column.created_at.DisplayAs.string")
      .last()
      .simulate("click");
  });
});
