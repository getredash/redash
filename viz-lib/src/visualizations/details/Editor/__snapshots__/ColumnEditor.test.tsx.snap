// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualizations -> Details -> Editor -> Column Editor Changes column alignment 1`] = `
Object {
  "alignContent": "right",
  "description": "",
  "displayAs": "number",
  "name": "amount",
  "title": "amount",
  "visible": true,
}
`;

exports[`Visualizations -> Details -> Editor -> Column Editor Changes column description 1`] = `
Object {
  "alignContent": "left",
  "description": "Current order status",
  "displayAs": "string",
  "name": "status",
  "title": "Status",
  "visible": true,
}
`;

exports[`Visualizations -> Details -> Editor -> Column Editor Changes column title 1`] = `
Object {
  "alignContent": "left",
  "description": "",
  "displayAs": "string",
  "name": "user_id",
  "title": "User ID",
  "visible": true,
}
`;

exports[`Visualizations -> Details -> Editor -> Column Editor Changes display type 1`] = `
Object {
  "alignContent": "left",
  "description": "",
  "displayAs": "string",
  "name": "created_at",
  "title": "Created At",
  "visible": true,
}
`;
